### PR TITLE
Bootstrap/restore bandwidth limit

### PR DESCRIPTION
**Pull request by Ben-PH**
* [ ] document all added functions
* [ ] try in sandbox /simulation/labnet
* [x] unit tests on the added/changed features
  * [x] make tests compile
  * [x] make tests pass 
* [ ] add logs allowing easy debugging in case the changes caused problems
* [ ] if the API has changed, update the API specification

### Summary
🚀🛠️🧪

This pull request adds bandwidth limiting to the bootstrap client and server using the `stream_limiter` crate. It changes the type of the `max_bytes_read_write` field in the bootstrap config structs and files from `f64` to `u64`, and updates the tests and port numbers accordingly.

> _To limit the bootstrap bandwidth use_
> _We wrap the TcpStream with a fuse_
> _We change the config type_
> _And fix some tests that hype_
> _The stream_limiter crate we introduce_

### Walkthrough
*  Add a dependency on the `stream_limiter` crate and use it to wrap `TcpStream` in the bootstrap client and server binders ([link](https://github.com/massalabs/massa/pull/4059/files?diff=unified&w=0#diff-bb9609263d51c8bedae0188499b1524875e2cd9101287c8d92042db31b36cc18R38), [link](https://github.com/massalabs/massa/pull/4059/files?diff=unified&w=0#diff-46782c0a3251886b3d3d654483bfda7302373961e336796f310646af2e13da59L21-R26), [link](https://github.com/massalabs/massa/pull/4059/files?diff=unified&w=0#diff-46782c0a3251886b3d3d654483bfda7302373961e336796f310646af2e13da59L45-R60), [link](https://github.com/massalabs/massa/pull/4059/files?diff=unified&w=0#diff-923a72f22a6d8df1ab1986c1911a4e163f8217df04be4dc5cb273665b9d562d6R27), [link](https://github.com/massalabs/massa/pull/4059/files?diff=unified&w=0#diff-923a72f22a6d8df1ab1986c1911a4e163f8217df04be4dc5cb273665b9d562d6L43-R44), [link](https://github.com/massalabs/massa/pull/4059/files?diff=unified&w=0#diff-923a72f22a6d8df1ab1986c1911a4e163f8217df04be4dc5cb273665b9d562d6L59-R65), [link](https://github.com/massalabs/massa/pull/4059/files?diff=unified&w=0#diff-923a72f22a6d8df1ab1986c1911a4e163f8217df04be4dc5cb273665b9d562d6R73-R82))
*  Add an optional limit parameter for the read and write bandwidth in bytes per second to the bootstrap client and server constructors and config structs ([link](https://github.com/massalabs/massa/pull/4059/files?diff=unified&w=0#diff-46782c0a3251886b3d3d654483bfda7302373961e336796f310646af2e13da59L175-R190), [link](https://github.com/massalabs/massa/pull/4059/files?diff=unified&w=0#diff-46782c0a3251886b3d3d654483bfda7302373961e336796f310646af2e13da59L207-R222), [link](https://github.com/massalabs/massa/pull/4059/files?diff=unified&w=0#diff-923a72f22a6d8df1ab1986c1911a4e163f8217df04be4dc5cb273665b9d562d6L95-R109), [link](https://github.com/massalabs/massa/pull/4059/files?diff=unified&w=0#diff-923a72f22a6d8df1ab1986c1911a4e163f8217df04be4dc5cb273665b9d562d6L214-R228), [link](https://github.com/massalabs/massa/pull/4059/files?diff=unified&w=0#diff-923a72f22a6d8df1ab1986c1911a4e163f8217df04be4dc5cb273665b9d562d6L316-R330), [link](https://github.com/massalabs/massa/pull/4059/files?diff=unified&w=0#diff-20eadb4ee40e0071fec6af0f8bcf8e8ce658f07335590ce21173b8099ad259c1R354), [link](https://github.com/massalabs/massa/pull/4059/files?diff=unified&w=0#diff-20eadb4ee40e0071fec6af0f8bcf8e8ce658f07335590ce21173b8099ad259c1R362), [link](https://github.com/massalabs/massa/pull/4059/files?diff=unified&w=0#diff-20eadb4ee40e0071fec6af0f8bcf8e8ce658f07335590ce21173b8099ad259c1R466), [link](https://github.com/massalabs/massa/pull/4059/files?diff=unified&w=0#diff-20eadb4ee40e0071fec6af0f8bcf8e8ce658f07335590ce21173b8099ad259c1L478-R512), [link](https://github.com/massalabs/massa/pull/4059/files?diff=unified&w=0#diff-2f2a0d279d4c015c632b27bdc7bd2fb62cf753fa8811dead19aadc3133666156R235), [link](https://github.com/massalabs/massa/pull/4059/files?diff=unified&w=0#diff-2f2a0d279d4c015c632b27bdc7bd2fb62cf753fa8811dead19aadc3133666156R252), [link](https://github.com/massalabs/massa/pull/4059/files?diff=unified&w=0#diff-455ddfa79024d518d5baefa18aaf0bff6d78763ab13eba505a0cc7a76e95a21bL65-R65), [link](https://github.com/massalabs/massa/pull/4059/files?diff=unified&w=0#diff-455ddfa79024d518d5baefa18aaf0bff6d78763ab13eba505a0cc7a76e95a21bL141-R141), [link](https://github.com/massalabs/massa/pull/4059/files?diff=unified&w=0#diff-455ddfa79024d518d5baefa18aaf0bff6d78763ab13eba505a0cc7a76e95a21bL154-R154), [link](https://github.com/massalabs/massa/pull/4059/files?diff=unified&w=0#diff-53705bdcd05cae716312f598a6d67411dc2242925b4194a2e77f391ea726d518L38-R44), [link](https://github.com/massalabs/massa/pull/4059/files?diff=unified&w=0#diff-53705bdcd05cae716312f598a6d67411dc2242925b4194a2e77f391ea726d518L84-R88), [link](https://github.com/massalabs/massa/pull/4059/files?diff=unified&w=0#diff-53705bdcd05cae716312f598a6d67411dc2242925b4194a2e77f391ea726d518R101-R102), [link](https://github.com/massalabs/massa/pull/4059/files?diff=unified&w=0#diff-53705bdcd05cae716312f598a6d67411dc2242925b4194a2e77f391ea726d518L132-R139), [link](https://github.com/massalabs/massa/pull/4059/files?diff=unified&w=0#diff-53705bdcd05cae716312f598a6d67411dc2242925b4194a2e77f391ea726d518L189-R196), [link](https://github.com/massalabs/massa/pull/4059/files?diff=unified&w=0#diff-53705bdcd05cae716312f598a6d67411dc2242925b4194a2e77f391ea726d518L234-R241), [link](https://github.com/massalabs/massa/pull/4059/files?diff=unified&w=0#diff-53705bdcd05cae716312f598a6d67411dc2242925b4194a2e77f391ea726d518R248), [link](https://github.com/massalabs/massa/pull/4059/files?diff=unified&w=0#diff-53705bdcd05cae716312f598a6d67411dc2242925b4194a2e77f391ea726d518L365-R373), [link](https://github.com/massalabs/massa/pull/4059/files?diff=unified&w=0#diff-53705bdcd05cae716312f598a6d67411dc2242925b4194a2e77f391ea726d518R380), [link](https://github.com/massalabs/massa/pull/4059/files?diff=unified&w=0#diff-8cd38f915cf6177aacea92ef29b97d8ceea3c69dc1d21ba95b55301b2ca258fdL396-R396), [link](https://github.com/massalabs/massa/pull/4059/files?diff=unified&w=0#diff-5275c26ef00940f5cb65a70bbaac1ffe9e40f1fd639b1bec4e7ae9f131972184L269-R269), [link](https://github.com/massalabs/massa/pull/4059/files?diff=unified&w=0#diff-1f883ba25a829313ad70daa6c31595706cf842a63e6e7bc0c342d89081084226L72-R72))
*  Access the underlying stream of the `Limiter` wrapper to set the read and write timeouts and transfer the message bytes ([link](https://github.com/massalabs/massa/pull/4059/files?diff=unified&w=0#diff-46782c0a3251886b3d3d654483bfda7302373961e336796f310646af2e13da59L175-R190), [link](https://github.com/massalabs/massa/pull/4059/files?diff=unified&w=0#diff-46782c0a3251886b3d3d654483bfda7302373961e336796f310646af2e13da59L207-R222), [link](https://github.com/massalabs/massa/pull/4059/files?diff=unified&w=0#diff-923a72f22a6d8df1ab1986c1911a4e163f8217df04be4dc5cb273665b9d562d6L95-R109), [link](https://github.com/massalabs/massa/pull/4059/files?diff=unified&w=0#diff-923a72f22a6d8df1ab1986c1911a4e163f8217df04be4dc5cb273665b9d562d6L214-R228), [link](https://github.com/massalabs/massa/pull/4059/files?diff=unified&w=0#diff-923a72f22a6d8df1ab1986c1911a4e163f8217df04be4dc5cb273665b9d562d6L316-R330))
*  Add a test function to check the bandwidth limit of the `Limiter` wrapper by sending and receiving long error messages ([link](https://github.com/massalabs/massa/pull/4059/files?diff=unified&w=0#diff-53705bdcd05cae716312f598a6d67411dc2242925b4194a2e77f391ea726d518R28), [link](https://github.com/massalabs/massa/pull/4059/files?diff=unified&w=0#diff-53705bdcd05cae716312f598a6d67411dc2242925b4194a2e77f391ea726d518L66-R71), [link](https://github.com/massalabs/massa/pull/4059/files?diff=unified&w=0#diff-53705bdcd05cae716312f598a6d67411dc2242925b4194a2e77f391ea726d518R95), [link](https://github.com/massalabs/massa/pull/4059/files?diff=unified&w=0#diff-53705bdcd05cae716312f598a6d67411dc2242925b4194a2e77f391ea726d518R101-R102), [link](https://github.com/massalabs/massa/pull/4059/files?diff=unified&w=0#diff-53705bdcd05cae716312f598a6d67411dc2242925b4194a2e77f391ea726d518L132-R139), [link](https://github.com/massalabs/massa/pull/4059/files?diff=unified&w=0#diff-53705bdcd05cae716312f598a6d67411dc2242925b4194a2e77f391ea726d518L189-R196), [link](https://github.com/massalabs/massa/pull/4059/files?diff=unified&w=0#diff-53705bdcd05cae716312f598a6d67411dc2242925b4194a2e77f391ea726d518R496-R603))
*  Change the port numbers of the protocol workers in the tests to avoid conflicts with the bootstrap ports ([link](https://github.com/massalabs/massa/pull/4059/files?diff=unified&w=0#diff-1dde26bed56ff3c99b9be1f786b58a8dab6b81cc9a9a2c697c5dba2dfe1e5af4L182-R182), [link](https://github.com/massalabs/massa/pull/4059/files?diff=unified&w=0#diff-1dde26bed56ff3c99b9be1f786b58a8dab6b81cc9a9a2c697c5dba2dfe1e5af4L194-R194), [link](https://github.com/massalabs/massa/pull/4059/files?diff=unified&w=0#diff-1dde26bed56ff3c99b9be1f786b58a8dab6b81cc9a9a2c697c5dba2dfe1e5af4L208-R208), [link](https://github.com/massalabs/massa/pull/4059/files?diff=unified&w=0#diff-1dde26bed56ff3c99b9be1f786b58a8dab6b81cc9a9a2c697c5dba2dfe1e5af4L221-R221))

### CODE REVIEW

File: massa-bootstrap/src/bindings/client.rs
Comment (Ben-PH) [inline massa-bootstrap/src/bindings/client.rs:54]: todo: remove me

File: massa-bootstrap/src/bindings/client.rs
Comment (Ben-PH) [inline massa-bootstrap/src/bindings/client.rs:61]: todo: comment me to provide context

File: massa-bootstrap/src/bindings/server.rs
Comment (Ben-PH) [inline massa-bootstrap/src/bindings/server.rs:81]: todo: comment for context

File: massa-bootstrap/src/client.rs
Comment (Ben-PH) [inline massa-bootstrap/src/client.rs:503]: todo: revert me

File: massa-bootstrap/src/client.rs
Comment (Ben-PH) [inline massa-bootstrap/src/client.rs:513]: todo: revert me

File: massa-bootstrap/src/server.rs
Comment (Ben-PH) [inline massa-bootstrap/src/server.rs:251]: this needs to be corrected...

File: massa-bootstrap/src/tests/binders.rs
Comment (Ben-PH) [inline massa-bootstrap/src/tests/binders.rs:497]: todo: update doc-comment and provide context comments in test

File: massa-protocol-worker/src/tests/mod.rs
Comment (Ben-PH) [inline massa-protocol-worker/src/tests/mod.rs:182]: This provided a fix locally, but fix doesn't work in CI

Comment (Ben-PH) [reply massa-protocol-worker/src/tests/mod.rs:182]: ...apparently it does?

Comment (Ben-PH) [reply massa-protocol-worker/src/tests/mod.rs:182]: ...this is weird. last time I looked at this page, the CI had passed.

**Comment by litchipi** [comment]
Note that for now `stream_limiter` is considered broken.
I'm working on fixing it (it fails when it comes to TcpStream limiting), but please do not merge this once everything is validated to be OK

**Comment by litchipi** [comment]
@Ben-PH You can now update `stream_limiter` to version `3.0.0` as the fix is finished

File: massa-bootstrap/src/tests/binders.rs
Comment (litchipi) [inline massa-bootstrap/src/tests/binders.rs:739]: Please note that this kind of assert may fail sometimes

That's because TCP packets are bound to 64Kb per packet, so even if we allow more data to be sent, we'll have to do multiple read.
Even without any limiting (sleep) occuring, this means that it's possible to be still slower than the rate limit set.

**Review by litchipi — CHANGES_REQUESTED** [review]

File: massa-bootstrap/src/tests/binders.rs
Comment (Ben-PH) [reply massa-bootstrap/src/tests/binders.rs:739]: That could possibly explain why CI is failing on windows.

Any suggestions for limit setup in tests, and how much of an assert window to provide?

File: massa-bootstrap/src/bindings/server.rs
Comment (Ben-PH) [reply massa-bootstrap/src/bindings/server.rs]: [a0582bc](https://github.com/massalabs/massa/pull/4059/commits/a0582bcebff978198fd06804d70ab326486fec62)

File: massa-bootstrap/src/bindings/client.rs
Comment (Ben-PH) [reply massa-bootstrap/src/bindings/client.rs]: [a0582bc](https://github.com/massalabs/massa/pull/4059/commits/a0582bcebff978198fd06804d70ab326486fec62)

**Review by litchipi — APPROVED** [review]
Let's try to make the CI green before merging

File: massa-bootstrap/src/tests/binders.rs
Comment (Ben-PH) [reply massa-bootstrap/src/tests/binders.rs:750]: yeah. Got things working with windows, but now mac is failing. takes about 20s. I'll account for mac, and assuming that's good, and green across the board, ill hit merge. Thanks!

**Pull request merged**
